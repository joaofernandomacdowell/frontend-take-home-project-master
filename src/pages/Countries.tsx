import React, { useState, useEffect } from "react";

import CountryCard from "../components/CountryCard";
import { Country } from "../api/types";

const ALL_COUNTRIES_ENDPOINT = "https://restcountries.eu/rest/v2/all";

const fetchCountryApiData = async (setter: Function) => {
  const response = await fetch(ALL_COUNTRIES_ENDPOINT);
  const countries = await response.json();
  setter(countries);
  console.log(countries);
};

const Countries = () => {
  const [countries, setCountries] = useState([]);

  useEffect(() => {
    fetchCountryApiData(setCountries);
  }, []);

  return (
    <>
      {countries.map((country: Country) => (
        <CountryCard key={country.name.toLowerCase()} country={country} />
      ))}
    </>
  );
};

export default Countries;
