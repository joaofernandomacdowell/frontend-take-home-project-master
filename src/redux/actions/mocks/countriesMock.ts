import { Country } from '../../../api/types';

const countriesMock: Country[] = [
  {
    name: 'American Samoa',
    alpha3Code: 'ASM',
    capital: 'Pago Pago',
    region: 'Oceania',
    subregion: 'Polynesia',
    population: 57100,
    nativeName: 'American Samoa',
    currencies: [
      {
        code: 'USD',
        name: 'United State Dollar',
        symbol: '$',
      },
    ],
    languages: [
      {
        iso639_1: 'en',
        iso639_2: 'eng',
        name: 'English',
        nativeName: 'English',
      },
      {
        iso639_1: 'sm',
        iso639_2: 'smo',
        name: 'Samoan',
        nativeName: "gagana fa'a Samoa",
      },
    ],
    flag: 'https://restcountries.eu/data/asm.svg',
  },
  {
    name: 'Cameroon',
    alpha3Code: 'CMR',
    capital: 'Yaoundé',
    region: 'Africa',
    subregion: 'Middle Africa',
    population: 22709892,
    nativeName: 'Cameroon',
    currencies: [
      {
        code: 'XAF',
        name: 'Central African CFA franc',
        symbol: 'Fr',
      },
    ],
    languages: [
      {
        iso639_1: 'en',
        iso639_2: 'eng',
        name: 'English',
        nativeName: 'English',
      },
      {
        iso639_1: 'fr',
        iso639_2: 'fra',
        name: 'French',
        nativeName: 'français',
      },
    ],
    flag: 'https://restcountries.eu/data/cmr.svg',
  },
  {
    name: 'United States of America',
    alpha3Code: 'USA',
    capital: 'Washington, D.C.',
    region: 'Americas',
    subregion: 'Northern America',
    population: 323947000,
    nativeName: 'United States',
    currencies: [
      {
        code: 'USD',
        name: 'United States dollar',
        symbol: '$',
      },
    ],
    languages: [
      {
        iso639_1: 'en',
        iso639_2: 'eng',
        name: 'English',
        nativeName: 'English',
      },
    ],
    flag: 'https://restcountries.eu/data/usa.svg',
  },
  {
    name: 'Virgin Islands (U.S.)',
    alpha3Code: 'VIR',
    capital: 'Charlotte Amalie',
    region: 'Americas',
    subregion: 'Caribbean',
    population: 114743,
    nativeName: 'Virgin Islands of the United States',
    currencies: [
      {
        code: 'USD',
        name: 'United States dollar',
        symbol: '$',
      },
    ],
    languages: [
      {
        iso639_1: 'en',
        iso639_2: 'eng',
        name: 'English',
        nativeName: 'English',
      },
    ],
    flag: 'https://restcountries.eu/data/vir.svg',
  },
];

export default countriesMock;
