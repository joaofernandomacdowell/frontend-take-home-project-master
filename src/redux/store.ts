/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { throttle } from 'lodash';

import { loadState, saveState } from '../localStorage/localStorage';

import rootReducer from './reducers';

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
const persistedState = loadState();

const store = createStore(
  rootReducer,
  persistedState,
  composeWithDevTools(applyMiddleware(thunk))
);

store.subscribe(
  throttle(() => {
    saveState({
      countriesState: store.getState().countriesState,
    });
  }, 1000)
);

export type AppState = ReturnType<typeof rootReducer>;

export default store;
