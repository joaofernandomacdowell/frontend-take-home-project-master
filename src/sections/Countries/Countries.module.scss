@import '../../styles/mixins';
@import '../../styles/variables.scss';

.container {
  display: grid;
  grid-template-columns: 1fr;
  gap: 40px;
  padding: 20px;

  @include media-small {
    grid-template-columns: repeat(2, 1fr);
  }

  @include media-medium {
    grid-template-columns: repeat(3, 1fr);
  }

  @include media-large {
    padding: 20px 100px;
    grid-template-columns: repeat(4, 1fr);
  }
}
