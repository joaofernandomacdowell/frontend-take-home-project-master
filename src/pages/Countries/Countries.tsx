import React, { useEffect, useState } from 'react';
import { connect, useDispatch } from 'react-redux';

import { AppState } from '../../redux/store';

import Pagination from '../../components/Pagination';
import Search from '../../components/Search';
import CountriesList from '../../components/CountriesList';
import { setSearchCountry } from '../../redux/actions/searchActions';
import { Country } from '../../api/types';

const ITEMS_PER_PAGE = 20;

interface CountriesProps {
  countries: Country[];
}

export const filterFn = (
  countries: Country[],
  text: string
): Country[] =>
  countries.filter((country) => {
    if (
      country.name.toLowerCase().indexOf(text.toLowerCase()) !== -1
    ) {
      return country;
    }

    return false;
  });

const Countries = ({ countries }: CountriesProps): JSX.Element => {
  const dispatch = useDispatch();

  const [currentPage, setCurrentPage] = useState(1);
  const [text, setText] = useState('');

  const indexOfLastCountry = currentPage * ITEMS_PER_PAGE;
  const indexOfFirstCountry = indexOfLastCountry - ITEMS_PER_PAGE;
  const currentCountries = countries.slice(
    indexOfFirstCountry,
    indexOfLastCountry
  );

  const [filtered, setFiltered] = useState(currentCountries);

  // Execute when input searchText has changed
  useEffect(() => {
    dispatch(setSearchCountry(text.trim()));
    setFiltered(filterFn(countries, text));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [text]);

  const paginate = (pageNumber: number) => setCurrentPage(pageNumber);

  return (
    <>
      <Search
        text={text}
        onChange={(searchTerm: string) => setText(searchTerm)}
      />
      <CountriesList
        countries={
          filtered.length === countries.length
            ? currentCountries
            : filtered
        }
      />
      <Pagination
        itemsPerPage={ITEMS_PER_PAGE}
        totalItems={filtered.length}
        paginate={paginate}
      />
    </>
  );
};

const mapStateToProps = ({ countriesState }: AppState) => {
  const { countries } = countriesState;

  return {
    countries,
  };
};

export default connect(mapStateToProps)(Countries);
