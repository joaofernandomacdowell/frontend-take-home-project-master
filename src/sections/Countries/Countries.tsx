import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import styles from './Countries.module.scss';

import Search from '../Search';

import CountryCard from '../../components/CountryCard';
import { Country } from '../../api/types';
import { fetchCountries } from '../../redux/actions/countriesActions';
import { AppState } from '../../redux/store';

interface CountriesProps {
  countries: Country[];
  fetchAllCountries: () => void;
  loading: boolean;
}

const Countries = ({
  loading,
  countries,
  fetchAllCountries,
}: CountriesProps): JSX.Element => {
  useEffect(() => {
    fetchAllCountries();
  }, [fetchAllCountries]);

  return loading && countries.length === 0 ? (
    <h2>Loading...</h2>
  ) : (
    <>
      <Search />
      <section className={styles.container}>
        {countries.map((country: Country) => (
          <CountryCard key={country.numericCode} country={country} />
        ))}
      </section>
    </>
  );
};

const mapStateToProps = (state: AppState) => ({
  loading: state.allCountries.isFetching,
  countries: state.allCountries.countries,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchAllCountries: fetchCountries(dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Countries);
