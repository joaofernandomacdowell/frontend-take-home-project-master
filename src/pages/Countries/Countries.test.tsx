/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from 'react';
import { Provider } from 'react-redux';

import { cleanup, render } from '@testing-library/react';
import Countries, { filterFn } from './Countries';
import countriesMock from '../../redux/actions/mocks/countriesMock';

import store from '../../redux/store';

afterEach(cleanup);

describe('Countries', () => {
  describe('snapshot', () => {
    test('Should match render snapshot', () => {
      const countries = render(
        <Provider store={store}>
          <Countries />
        </Provider>
      );
      expect(countries).toMatchSnapshot();
    });
  });

  describe('functions', () => {
    test('filterFn returns correct filtered array', () => {
      const countries = countriesMock;
      const text = 'amo';

      const expectedFilter = filterFn(countries, text);

      // match country American Samoa
      const shouldMatchOnly = [countriesMock[0]];

      expect(expectedFilter).toStrictEqual(shouldMatchOnly);
    });
  });
});
