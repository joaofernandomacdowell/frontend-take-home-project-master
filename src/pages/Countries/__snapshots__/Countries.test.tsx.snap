// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries Should match render snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container "
      >
        <div
          class="row "
        >
          <section
            class="search"
          >
            <input
              class="searchInput"
              id="search"
              name="search"
              placeholder="Search for a country"
              type="search"
              value=""
            />
          </section>
        </div>
      </div>
      <div
        class="container"
      >
        <span
          class="message"
        >
          No country was found
        </span>
        
      </div>
      <nav>
        <ul
          class="pagination"
        />
      </nav>
    </div>
  </body>,
  "container": <div>
    <div
      class="container "
    >
      <div
        class="row "
      >
        <section
          class="search"
        >
          <input
            class="searchInput"
            id="search"
            name="search"
            placeholder="Search for a country"
            type="search"
            value=""
          />
        </section>
      </div>
    </div>
    <div
      class="container"
    >
      <span
        class="message"
      >
        No country was found
      </span>
      
    </div>
    <nav>
      <ul
        class="pagination"
      />
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Countries snapshot Should match render snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container "
      >
        <div
          class="row "
        >
          <section
            class="search"
          >
            <input
              class="searchInput"
              id="search"
              name="search"
              placeholder="Search for a country"
              type="search"
              value=""
            />
          </section>
        </div>
      </div>
      <div
        class="container"
      >
        <span
          class="message"
        >
          No country was found
        </span>
        
      </div>
      <nav>
        <ul
          class="pagination"
        />
      </nav>
    </div>
  </body>,
  "container": <div>
    <div
      class="container "
    >
      <div
        class="row "
      >
        <section
          class="search"
        >
          <input
            class="searchInput"
            id="search"
            name="search"
            placeholder="Search for a country"
            type="search"
            value=""
          />
        </section>
      </div>
    </div>
    <div
      class="container"
    >
      <span
        class="message"
      >
        No country was found
      </span>
      
    </div>
    <nav>
      <ul
        class="pagination"
      />
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
